src/host/VtIo.cpp

```diff
@@ -373,148 +267,392 @@ void VtIo::SetWindowVisibility(bool showOrHide) noexcept
-// Arguments:
-// - <none>
-// Return Value:
-// - S_OK if the renderer successfully suppressed the next repaint, otherwise an
-//      appropriate HRESULT indicating failure.
-[[nodiscard]] HRESULT VtIo::SuppressResizeRepaint()
+void VtIo::CloseInput()
+{
+    _pVtInputThread = nullptr;
+    SendCloseEvent();
+}
+
+void VtIo::CloseOutput()
 {
-    auto hr = S_OK;
-    if (_pVtRenderEngine)
+    _hOutput.reset();
+}
+
+void VtIo::SendCloseEvent()
+{
+    LockConsole();
+    const auto unlock = wil::scope_exit([] { UnlockConsole(); });
+
+    // This function is called when the ConPTY signal pipe is closed (PtySignalInputThread) and when the input
+    // pipe is closed (VtIo). Usually these two happen at about the same time. This if condition is a bit of
+    // a premature optimization and prevents us from sending out a CTRL_CLOSE_EVENT right after another.
+    if (!std::exchange(_closeEventSent, true))
     {
-        hr = _pVtRenderEngine->SuppressResizeRepaint();
+        CloseConsoleProcessState();
     }
-    return hr;
 }
 
-// Method Description:
-// - Attempts to set the initial cursor position, if we're looking for it.
-//      If we're not trying to inherit the cursor, does nothing.
-// Arguments:
-// - coordCursor: The initial position of the cursor.
-// Return Value:
-// - S_OK if we successfully inherited the cursor or did nothing, else an
-//      appropriate HRESULT
-[[nodiscard]] HRESULT VtIo::SetCursorPosition(const til::point coordCursor)
+// Returns true for C0 characters and C1 [single-character] CSI.
+// A copy of isActionableFromGround() from stateMachine.cpp.
+bool VtIo::IsControlCharacter(wchar_t wch) noexcept
 {
-    auto hr = S_OK;
-    if (_lookingForCursorPosition)
+    // This is equivalent to:
+    //   return (wch <= 0x1f) || (wch >= 0x7f && wch <= 0x9f);
+    // It's written like this to get MSVC to emit optimal assembly for findActionableFromGround.
+    // It lacks the ability to turn boolean operators into binary operations and also happens
+    // to fail to optimize the printable-ASCII range check into a subtraction & comparison.
+    return (wch <= 0x1f) | (static_cast<wchar_t>(wch - 0x7f) <= 0x20);
+}
+
+static size_t formatAttributes(char (&buffer)[16], WORD attributes) noexcept
+{
+    auto end = &buffer[0];
+    memcpy(end, "\x1b[0", 4);
+    end += 3;
+
+    if (attributes & COMMON_LVB_REVERSE_VIDEO)
     {
-        if (_pVtRenderEngine)
-        {
-            hr = _pVtRenderEngine->InheritCursor(coordCursor);
-        }
+        memcpy(end, ";7", 2);
+        end += 2;
+    }
 
-        _lookingForCursorPosition = false;
+    // `attributes` of exactly `FOREGROUND_BLUE | FOREGROUND_GREEN | FOREGROUND_RED`
+    // are often used to indicate the default colors in Windows Console applications.
+    // Since we always emit SGR 0 (reset all attributes), we simply need to skip this branch.
+    if ((attributes & (FG_ATTRS | BG_ATTRS)) != (FOREGROUND_BLUE | FOREGROUND_GREEN | FOREGROUND_RED))

```

Comment:
I'll use `TextAttribute` for now but I think this will change in the future.

My intention is for the server project to be a slim standalone library at some point in the near (?) future. I felt like it'd be easier to cleanly split it out of the rest of the project if we didn't rely on widely used classes that may accidentally pull in larger dependencies. Basically, I'd love if the console server was something along the lines of "3 files and 2000 lines of code". (Not meant literally - just to give you a picture. But it's probably actually a realistic goal.)